src/core/bpf-pinned-prog.c

```diff
@@ -0,0 +1,93 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "bpf-pinned-prog.h"
+#include "cgroup-bpf-ctx.h"
+
+#include "unit.h"
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(bpf_prog_hash_ops, void, trivial_hash_func, trivial_compare_func, BPFProgram, bpf_program_unref);
+
+static int load_programs(Unit *u, const CGroupContext *c, Hashmap *progs_by_id) {
+        CGroupPinnedBPFProgram *p;
+        int r;
+
+        assert(u);
+        assert(c);
+        assert(progs_by_id);
+
+        LIST_FOREACH(prog, p, c->pinned_bpf_progs) {
+                _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+
+                r = bpf_program_new_from_bpffs_path(p->bpffs_path, p->attach_type, &prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Loading of BPF program %s failed: %m", p->bpffs_path);
+
+                r = hashmap_put(progs_by_id, p, TAKE_PTR(prog));
```

Comment:
what happens to "prog" if hashmap_put fails?

Reply:
@bluca 

Nice catch, thanks! Rewrote this snippet as
```
                r = hashmap_put(progs_by_id, p, prog);
                if (r < 0)
                        return log_unit_error_errno(u, r, "Can't put program to hashmap: %m");

                TAKE_PTR(prog);
```